### PR TITLE
schedule: use change peer v2 for single operator step if possible

### LINKED ISSUE #5407: Use change peer v2 for single operator step
## Development Task

Precedent work of https://github.com/tikv/tikv/issues/12876. Witness would be only implemented in v2. To support witness, PD should do single operator step also by conf change v2.

### PR DESCRIPTION
**Pull request by Connor1996**
### What problem does this PR solve?

Issue Number: Close #5407

### What is changed and how does it work?

```commit-message
Use change peer v2 for single operator step if possible
```

### Check List

Tests

- No code

### Release note

```release-note
Use change peer v2 for single operator step if possible
```

### CODE REVIEW

File: server/schedule/operator/step.go [hunk 1]
```diff
@@ -763,3 +845,48 @@ func fastStepWaitDuration(regionSize int64) time.Duration {
...
+		Peer: &metapb.Peer{
+			Id:      id,
+			StoreId: storeID,
+			Role:    metapb.PeerRole_Voter,
+		},
+	}
+	return createResponse(change, useConfChangeV2)
+}
+
+func addLearnerNode(id, storeID uint64, useConfChangeV2 bool) *pdpb.RegionHeartbeatResponse {
```
Comment (rleungx) [inline server/schedule/operator/step.go:861]: How about removing these two functions and directly putting the inside logic into GetCmd?

File: server/schedule/operator/step.go [hunk 2]
```diff
@@ -266,7 +300,7 @@ func (pl PromoteLearner) IsFinish(region *core.RegionInfo) bool {
 // CheckInProgress checks if the step is in the progress of advancing.
 func (pl PromoteLearner) CheckInProgress(_ ClusterInformer, region *core.RegionInfo) error {
 	peer := region.GetStorePeer(pl.ToStore)
-	if peer.GetId() != pl.PeerID {
+	if peer == nil || peer.GetId() != pl.PeerID {
```
Comment (rleungx) [inline server/schedule/operator/step.go:303]: Why change it?

Comment (Connor1996) [reply server/schedule/operator/step.go:303]: more robust

File: server/schedule/operator/step.go (hunk 1 above)
Comment (Connor1996) [reply server/schedule/operator/step.go:861]: it's reused by too steps

**Review by HunDunDM — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

File: server/schedule/operator/step.go [hunk 3]
```diff
@@ -576,40 +658,32 @@ func (cpe ChangePeerV2Enter) CheckInProgress(_ ClusterInformer, region *core.Reg
...
 	for _, pl := range cpe.PromoteLearners {
-		changes = append(changes, &pdpb.ChangePeer{
-			ChangeType: eraftpb.ConfChangeType_AddNode,
-			Peer: &metapb.Peer{
-				Id:      pl.PeerID,
-				StoreId: pl.ToStore,
-				Role:    metapb.PeerRole_Voter,
-			},
-		})
+		changes = append(changes, pl.GetCmd(region, useConfChangeV2).ChangePeer)
```
Comment (ethercflow) [inline server/schedule/operator/step.go:675]:
```suggestion
		changes = append(changes, pl.GetCmd(region, useConfChangeV2).ChangePeerV2.Changes...)
```

File: server/schedule/operator/step.go [hunk 4]
```diff
@@ -576,40 +658,32 @@ func (cpe ChangePeerV2Enter) CheckInProgress(_ ClusterInformer, region *core.Reg
...
 	for _, dv := range cpe.DemoteVoters {
-		changes = append(changes, &pdpb.ChangePeer{
-			ChangeType: eraftpb.ConfChangeType_AddLearnerNode,
-			Peer: &metapb.Peer{
-				Id:      dv.PeerID,
-				StoreId: dv.ToStore,
-				Role:    metapb.PeerRole_Learner,
-			},
-		})
+		changes = append(changes, dv.GetCmd(region, useConfChangeV2).ChangePeer)
```
Comment (ethercflow) [inline server/schedule/operator/step.go:678]:
```suggestion
		changes = append(changes, dv.GetCmd(region, useConfChangeV2).ChangePeerV2.Changes...)
```

File: server/schedule/operator/step.go (hunk 3 above)
Comment (Connor1996) [reply server/schedule/operator/step.go:675]: nice catch

File: server/schedule/operator/step.go [hunk 5]
```diff
@@ -173,6 +188,16 @@ func (ap AddPeer) Timeout(start time.Time, regionSize int64) bool {
...
 }
 
+// GetCmd returns the schedule command for heartbeat response.
+func (ap AddPeer) GetCmd(region *core.RegionInfo, useConfChangeV2 bool) *pdpb.RegionHeartbeatResponse {
+	peer := region.GetStorePeer(ap.ToStore)
+	if peer != nil {
+		// The newly added peer is pending.
+		return nil
+	}
+	return addNode(ap.PeerID, ap.ToStore, useConfChangeV2)
```
Comment (HunDunDM) [inline server/schedule/operator/step.go:198]: How about `createResponse(addNode(ap.PeerID, ap.ToStore), useConfChangeV2)`?

File: server/schedule/operator/step.go [hunk 6]
```diff
@@ -576,40 +658,32 @@ func (cpe ChangePeerV2Enter) CheckInProgress(_ ClusterInformer, region *core.Reg
...
 	for _, pl := range cpe.PromoteLearners {
-		changes = append(changes, &pdpb.ChangePeer{
-			ChangeType: eraftpb.ConfChangeType_AddNode,
-			Peer: &metapb.Peer{
-				Id:      pl.PeerID,
-				StoreId: pl.ToStore,
-				Role:    metapb.PeerRole_Voter,
-			},
-		})
+		changes = append(changes, pl.GetCmd(region, useConfChangeV2).ChangePeerV2.Changes...)
```
Comment (HunDunDM) [inline server/schedule/operator/step.go:675]: How about `changes = append(changes, addLearnerNode(pl.PeerID, pl.ToStore))`

File: server/schedule/operator/step.go [hunk 7]
```diff
@@ -463,6 +536,15 @@ func (dv DemoteVoter) Timeout(start time.Time, regionSize int64) bool {
 	return time.Since(start) > fastStepWaitDuration(regionSize)
 }
 
+// GetCmd returns the schedule command for heartbeat response.
+func (dv DemoteVoter) GetCmd(region *core.RegionInfo, useConfChangeV2 bool) *pdpb.RegionHeartbeatResponse {
```
Comment (HunDunDM) [inline server/schedule/operator/step.go:540]: `DemoteVoter` is not an `OpStep`. We don't need to implement this function.

Comment (Connor1996) [reply server/schedule/operator/step.go:540]: To make it consistent with others, we can simply call `GetCmd()` of it in `ChangePeerV2Enter`

**Pull request merged**